node/pkg/governor/governor.go

```diff
@@ -688,3 +715,54 @@ func (gov *ChainGovernor) HashFromMsg(msg *common.MessagePublication) string {
 	digest := v.SigningDigest()
 	return hex.EncodeToString(digest.Bytes())
 }
+
+// previouslyApproved checks to see if this VAA has previously reached quorum and we are listed as a signer. If so, the VAA can be allowed to continue.
+func (gov *ChainGovernor) previouslyApproved(msg *common.MessagePublication, newDigest string) (bool, error) {
+	// Look up the VAA in the database.
+	vaaID := db.VAAID{
+		EmitterChain:   msg.EmitterChain,
+		EmitterAddress: msg.EmitterAddress,
+		Sequence:       msg.Sequence,
+	}
+	vb, err := gov.db.GetSignedVAABytes(vaaID)
+	if err != nil {
+		if err != db.ErrVAANotFound {
+			return false, fmt.Errorf(`failed to look up VAA "%s" in the database: %v`, msg.MessageIDString(), err)
+		}
+		return false, nil
+	}
+
+	v, err := vaa.Unmarshal(vb)
+	if err != nil {
+		panic(fmt.Sprintf(`cgov: failed to unmarshal VAA "%s" from db: %v`, msg.MessageIDString(), err))
+	}
+
+	// Verify the digest matches.
+	oldDigest := hex.EncodeToString(v.SigningDigest().Bytes())
+	if oldDigest != newDigest {
+		gov.logger.Warn("unable to approve previously stored VAA because the digest has changed",
+			zap.String("msgID", msg.MessageIDString()),
+			zap.String("oldDigest", oldDigest),
+			zap.String("newDigest", newDigest),
+		)
+		return false, nil
+	}
+
+	// Generate our signature for this VAA so we can see if we are listed as a signer.
+	sig, err := crypto.Sign(v.SigningDigest().Bytes(), gov.gk)
```

Comment:
Right here I was expecting that we look through `v.Signatures` to find our index, then recover the public key from that signature (given that we just verified the hashes match above). If the public key matches our guardian's public key, we should be good to proceed. That would avoid needing the private key here, which would be best avoided.

Reply:
I like this approach much better! I have changed it to use `Ecrecover` with the public key rather than the private key.